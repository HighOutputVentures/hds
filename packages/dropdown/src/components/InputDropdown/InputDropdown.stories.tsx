import { Flex } from '@chakra-ui/react';
import ThemeProvider from '@highoutput/hds/src/components/ThemeProvider';
import { useArgs } from '@storybook/addons';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';
import SearchIcon from '../Icon/SearchIcon';
import InputDropdown from './InputDropdown';

export default {
  title: 'Components/Input Dropdown',
  component: InputDropdown,
} as ComponentMeta<typeof InputDropdown>;

const Template: ComponentStory<typeof InputDropdown> = (props) => {
  const [args, setArgs] = useArgs();

  const onChange = (value: any[]) => {
    setArgs({ ...args, value });
  };

  return (
    <ThemeProvider>
      <Flex gap={4} flexDir="column">
        <InputDropdown
          {...props}
          onChangeValue={onChange}
          options={args.options}
          helperMsg="This is a hint to help user"
          inputLeftIcon={SearchIcon}
        />
      </Flex>
    </ThemeProvider>
  );
};

export const Default = Template.bind({});

const options = [
  {
    value: 'Phoenix',
    label: 'Phoenix',
    username: '@phoenix',
    avatar:
      'https://assets.teenvogue.com/photos/626abe370979f2c5ace0ab29/16:9/w_2560%2Cc_limit/GettyImages-1352932505.jpg',
  },
  {
    value: 'Olivia',
    label: 'Olivia',
    username: '@olivia',
    avatar:
      'https://static01.nyt.com/images/2021/05/21/arts/21review-rodrigo01/merlin_188054001_2a34e77d-e653-488a-a4d9-1a0f1ddf73e4-superJumbo.jpg',
  },
];

Default.args = {
  ...Default.args,
  label: 'Users',
  options: options,
  placeholder: 'Add users',
};
