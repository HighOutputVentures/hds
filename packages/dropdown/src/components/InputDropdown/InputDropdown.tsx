import { ChevronDownIcon, ChevronUpIcon } from '@chakra-ui/icons';
import { As, Avatar, Flex, HStack, Icon, Text } from '@chakra-ui/react';
// @ts-ignore
import { FormContainer, FormContainerProps } from '@highoutput/hds-forms';
import {
  chakraComponents,
  MultiValue,
  Select,
  SingleValue,
} from 'chakra-react-select';
import React from 'react';
// @ts-ignore
import { CheckIcon } from '@highoutput/hds-icons';
import getStyles from './styles';
export interface Item {
  value: string | number;
  label: string;
}

export interface OptionItem extends Item {
  avatar?: string;
  username?: string;
}

export interface InputDropdownFieldProps
  extends Omit<FormContainerProps, 'partProps'> {
  options: OptionItem[];
  loading?: boolean;
  // fieldLabelProps?: Omit<BoxProps, 'children'>;
  required?: boolean;
  placement?: 'auto' | 'top' | 'bottom';
  multiple?: boolean;

  darkMode?: boolean;
  autoFocus?: boolean;
  inputLeftIcon?: As<any> | undefined;
  placeholder?: string;
  // partProps?: Partial<AutoCompleteInputFieldPartProps>;
  value?: string | string[] | number | number[];
  onChangeValue: (...event: any[]) => void;
  __selectTestId?: string;
}

const InputDropdown = (props: InputDropdownFieldProps) => {
  const {
    value,
    options,
    darkMode,
    placement,
    inputLeftIcon,
    errorMsg,
    autoFocus,
    disabled,
    __selectTestId,
    onChangeValue,
    multiple,
    loading,
    placeholder,
  } = props;

  const [optionVal, setOptionVal] = React.useState(options);

  const styles = getStyles({
    error: Boolean(errorMsg),
    multiple,
    darkMode,
  });
  const handleInputChange = (value: string) => {
    if (value.length > 1) {
      setOptionVal(
        options.filter((d) =>
          d.label.toLowerCase().includes(value.toLowerCase())
        )
      );
    } else setOptionVal(options);
  };

  return (
    <FormContainer {...props}>
      <Flex borderRadius={'8px'} position="relative" gap="10px">
        <Flex
          align={'center'}
          justify={'center'}
          w="20px"
          height={'100%'}
          position="absolute"
          left="14px"
          zIndex={1}
        >
          <Icon as={inputLeftIcon} w="24px" height="24px" />
        </Flex>

        <Select
          options={options}
          chakraStyles={styles}
          placeholder={placeholder}
          menuPlacement={placement ?? 'auto'}
          isMulti={true}
          inputId="input-dropdown"
          useBasicStyles
          aria-label="input-dropdown"
          captureMenuScroll
          backspaceRemovesValue
          autoFocus={autoFocus}
          isLoading={loading}
          data-testid={__selectTestId ?? `hds.input.dropdown.field`}
          isSearchable
          onInputChange={(e, _) => handleInputChange(e)}
          isDisabled={disabled}
          isClearable
          components={{
            ClearIndicator: () => null,
            IndicatorSeparator: () => null,
            DropdownIndicator: ({ selectProps }) => {
              const icon = selectProps.menuIsOpen
                ? ChevronUpIcon
                : ChevronDownIcon;
              return <Icon as={icon} w={4} h={6} stroke="brand.primary.500" />;
            },

            MultiValueContainer: ({ children, data, ...props }) => (
              <chakraComponents.MultiValueContainer {...props} data={data}>
                <HStack align="center" spacing="5px">
                  {data.avatar && (
                    <Avatar width="16px" height="16px" src={data.avatar} />
                  )}
                  {children}
                </HStack>
              </chakraComponents.MultiValueContainer>
            ),

            MenuList: ({ getValue, ...props }) => (
              <chakraComponents.MenuList getValue={getValue} {...props}>
                {optionVal.map((v) => {
                  return (
                    <Flex
                      onClick={() =>
                        props.selectOption({
                          value: v.value,
                          label: v.label,
                        })
                      }
                      cursor="pointer"
                      justify={'space-between'}
                      align="center"
                      py={'10px'}
                      px={'14px'}
                    >
                      <HStack align="center" spacing="8px">
                        <Avatar width="24px" height="24px" src={v.avatar} />

                        <Text fontSize={'18px'}>{v.label}</Text>
                        <Text fontSize={'18px'} color="neutrals.500">
                          {v.username}
                        </Text>
                      </HStack>
                      {getValue().map((d) => {
                        return d.label === v.label ? (
                          <Icon as={CheckIcon} />
                        ) : null;
                      })}
                    </Flex>
                  );
                })}
              </chakraComponents.MenuList>
            ),
          }}
          onChange={(
            options: MultiValue<Item> | (SingleValue<Item> | null)
          ) => {
            return isArray<MultiValue<Item>>(options)
              ? onChangeValue(options.map((o) => o.value))
              : onChangeValue(options?.value ?? null);
          }}
          value={options.filter((option) => {
            return Array.isArray(value)
              ? value.some((val: string | number) => val === option.value)
              : value === option.value;
          })}
        />
      </Flex>
    </FormContainer>
  );
};

export default InputDropdown;

function isArray<T>(subject: unknown): subject is T {
  return Array.isArray(subject);
}
