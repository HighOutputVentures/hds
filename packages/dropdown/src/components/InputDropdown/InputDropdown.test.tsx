import { ChakraProvider } from '@chakra-ui/react';
import { render, screen } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import InputDropdown from './InputDropdown';
const OPTIONS = ['user1', 'user2', 'user3', 'user4', 'user5'];

const SELECT_OPTION = Object.entries(OPTIONS).map(([value, label]) => {
  return {
    value,
    label,
  };
});

describe('Input Dropdown Component', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });
  beforeEach(() => {
    render(
      <ChakraProvider>
        <InputDropdown
          label="Users"
          options={SELECT_OPTION}
          __selectTestId={'hds.input.field'}
          __formLabelTestId={'hds.form.control.label'}
          onChangeValue={jest.fn()}
        />
      </ChakraProvider>
    );
  });
  it('Should render Input Dropdown  form container', async () => {
    const formControl = await screen.findAllByTestId('hds.form.control');
    expect(formControl).toHaveLength(1);
  });
  it('Should render Input Dropdown  label', async () => {
    const formControlLabel = await screen.findAllByTestId(
      'hds.form.control.label'
    );
    expect(formControlLabel).toHaveLength(1);
  });
  it('Should render Input Dropdown ', async () => {
    const inputDropdownField = await screen.findAllByRole('combobox', {
      name: 'input-dropdown',
    });

    expect(inputDropdownField).toHaveLength(1);
  });
});
describe('Snapshot', () => {
  it('should match snapshot', () => {
    const component = renderer.create(
      <ChakraProvider>
        <InputDropdown
          label="Users"
          options={SELECT_OPTION}
          onChangeValue={jest.fn()}
        />
      </ChakraProvider>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    component.unmount();
  });
});
