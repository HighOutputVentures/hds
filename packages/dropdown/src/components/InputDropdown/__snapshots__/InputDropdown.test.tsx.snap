// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot should match snapshot 1`] = `
Array [
  <div
    className="chakra-form-control css-1kxonj9"
    data-testid="hds.form-group.group"
    role="group"
  >
    <label
      className="chakra-form__label css-dh7ao0"
      data-testid="hds.form-group.label"
      htmlFor="field-:r0:"
      id="field-:r0:-label"
    >
      Users
    </label>
    <div
      className="css-1kqykvk"
    >
      <div
        className="css-10swtys"
      >
        <svg
          className="chakra-icon css-n059si"
          focusable={false}
          viewBox="0 0 24 24"
        >
          <g
            stroke="currentColor"
            strokeWidth="1.5"
          >
            <path
              d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
              fill="none"
              strokeLinecap="round"
            />
            <path
              d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
              fill="currentColor"
              strokeLinecap="round"
            />
            <circle
              cx="12"
              cy="12"
              fill="none"
              r="11.25"
              strokeMiterlimit="10"
            />
          </g>
        </svg>
      </div>
      <div
        className=" css-1iu7u4f"
        onKeyDown={[Function]}
      >
        <span
          className="css-1f43avz-a11yText-A11yText"
          id="react-select-5-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          className="css-1f43avz-a11yText-A11yText"
        />
        <div
          className=" css-citx3a"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <div
            className=" css-sy0cn0"
          >
            <div
              className=" css-1puyje9"
              id="react-select-5-placeholder"
            >
              Select...
            </div>
            <div
              className=" css-svnjlw"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-5-placeholder"
                aria-expanded={false}
                aria-haspopup={true}
                aria-label="input-dropdown"
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect="off"
                className=" css-146k1ar-input"
                disabled={false}
                id="input-dropdown"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                role="combobox"
                spellCheck="false"
                tabIndex={0}
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className=" css-hfbj6y"
          >
            <svg
              className="chakra-icon chakra-icon css-fk2u4o"
              focusable={false}
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <span />,
]
`;
